src/resources.js

```diff
@@ -475,22 +530,64 @@ export class Resources {
       let minTop = -1;
       for (let i = 0; i < changeHeightRequests.length; i++) {
         let request = changeHeightRequests[i];
+        let resource = request.resource;
         let box = request.resource.getLayoutBox();
         if (box.height == request.newHeight) {
           // Nothing to do.
           continue;
         }
-        if (box.top >= 0) {
-          minTop = minTop == -1 ? box.top : Math.min(minTop, box.top);
+
+        // Check resize rules. It will either resize element immediately, or
+        // wait until scrolling stops or will call the fallback.
+        let resize = false;
+        if (request.force || !this.visible_) {
+          // 1. An immediate execution requested or the document is hidden.
+          resize = true;
+        } else if (activeElement && (activeElement == resource.element ||
+              resource.element.contains(activeElement))) {
+          // 2. Active elements are immediately resized. The assumption is that
+          //    the resize is triggered by the user action or soon after.
+          resize = true;
+        } else if (box.bottom >= viewportRect.bottom - offset) {
+          // 3. Elements under viewport are resized immediately.
+          resize = true;
+        } else if (box.bottom <= viewportRect.top + offset) {
+          // 4. Elements above the viewport can only be resized when scrolling
+          //    has stopped, otherwise defer util next cycle.
+          if (isScrollingStopped) {
+            resize = true;
+          } else {
+            // Defer till next cycle.
+            this.changeHeightRequests_.push(request);
+          }
+        } else if (request.newHeight < box.height) {
+          // 5. Resize immediately if the requested size is smaller.
```

Comment:
With some tests it looked horrible - basically just white space for no good reason. It'd be wrong to show "overflow" element in this case as well - it's not what it means. I expect "overflow" elements to say something like "read more", not less. Showing overflowing component also looks bad - content then bunch of white space and then "overflow" element. So, as the result, it looks very bad and we can't give a user a user-action to correct it. Thus I decided to go ahead and do the resize.

However, there's a silver lining:
1. Reducing size should not be a dominant case. And when it is - it's likely because of a use action.
2. I still force it to follow scrolling rules, so it will not cause much of a jank.